src/components/stream-source/component/StreamGenerator.ts

```diff
@@ -0,0 +1,77 @@
+interface DeferredPrimise {
+  promise?: Promise<unknown>;
+  resolve?(value?: unknown): void;
+  reject?(): void;
+}
+
+enum StreamStatus {
+  ACTIVE,
+  CLOSED,
+}
+
+class StreamGenerator {
+  private _next: DeferredPrimise = StreamGenerator.createDeferredPromise();
+  private _status: StreamStatus = StreamStatus.ACTIVE;
+  private _buffer: string[] = [];
+  private _err?: Error;
+  private _stream: AsyncGenerator<string>;
+  private _loadingSignal: DeferredPrimise =
+    StreamGenerator.createDeferredPromise();
+
+  private static createDeferredPromise() {
+    const deferred: DeferredPrimise = {};
+    deferred.promise = new Promise((_resolve, _reject) => {
+      deferred.resolve = _resolve;
+      deferred.reject = _reject;
+    });
+    return deferred;
+  }
+
+  private async *generator() {
+    while (this.status === StreamStatus.ACTIVE) {
+      await this._next.promise;
+
+      this._next = StreamGenerator.createDeferredPromise();
+
+      if (this._err) {
+        throw this._err;
+      }
+
+      while (this._buffer.length) {
+        yield this._buffer.shift() as string;
```

Comment:
Not sure it's necessary to loop here, think it could be simplified to:

```
if (this._buffer.length) {
  const chunk = this._buffer.join("");
  this._buffer.length = 0;
  yield chunk;
}
```